[Tray] Does not work properly with other tray hosts // Can't install using `cargo`, and didn't start after building from source

**Describe the bug**
> A clear and concise description of what the bug is.

Due to `ironbar` not available on distro repos, I tried to install with `cargo install ironbar` but didn't compile

```
error[E0432]: unresolved import `nix::fcntl::OFlag`
 --> /home/salo/.cargo/registry/src/index.crates.io-6f17d22bba15001f/ironbar-0.15.1/src/clients/wayland/wlr_data_control/offer.rs:3:5
  |
3 | use nix::fcntl::OFlag;
  |     ^^^^^^^^^^^^^^^^^ no `OFlag` in `fcntl`

error[E0432]: unresolved import `nix::unistd::pipe2`
    --> /home/salo/.cargo/registry/src/index.crates.io-6f17d22bba15001f/ironbar-0.15.1/src/clients/wayland/wlr_data_control/offer.rs:4:26
     |
4    | use nix::unistd::{close, pipe2};
     |                          ^^^^^
     |                          |
     |                          no `pipe2` in `unistd`
     |                          help: a similar name exists in the module: `pipe`
     |
note: found an item that was configured out
    --> /home/salo/.cargo/registry/src/index.crates.io-6f17d22bba15001f/nix-0.27.1/src/unistd.rs:1233:8
     |
1233 | pub fn pipe2(flags: OFlag) -> Result<(RawFd, RawFd)> {
     |        ^^^^^
     = note: the item is gated behind the `fs` feature

error[E0432]: unresolved imports `nix::fcntl::fcntl`, `nix::fcntl::F_GETPIPE_SZ`, `nix::fcntl::F_SETPIPE_SZ`
   --> /home/salo/.cargo/registry/src/index.crates.io-6f17d22bba15001f/ironbar-0.15.1/src/clients/wayland/wlr_data_control/mod.rs:13:18
    |
13  | use nix::fcntl::{fcntl, F_GETPIPE_SZ, F_SETPIPE_SZ};
    |                  ^^^^^  ^^^^^^^^^^^^  ^^^^^^^^^^^^ no `F_SETPIPE_SZ` in `fcntl`
    |                  |      |
    |                  |      no `F_GETPIPE_SZ` in `fcntl`
    |                  no `fcntl` in `fcntl`
    |
    = help: consider importing this module instead:
            nix::fcntl
note: found an item that was configured out
   --> /home/salo/.cargo/registry/src/index.crates.io-6f17d22bba15001f/nix-0.27.1/src/fcntl.rs:509:8
    |
509 | pub fn fcntl(fd: RawFd, arg: FcntlArg) -> Result<c_int> {
    |        ^^^^^
    = note: the item is gated behind the `fs` feature

error[E0308]: mismatched types
   --> /home/salo/.cargo/registry/src/index.crates.io-6f17d22bba15001f/ironbar-0.15.1/src/clients/compositor/hyprland.rs:156:51
    |
156 |                     let name = get_workspace_name(workspace_type);
    |                                ------------------ ^^^^^^^^^^^^^^ expected `WorkspaceType`, found `WorkspaceDestroyedEventData`
    |                                |
    |                                arguments to this function are incorrect
    |
note: function defined here
   --> /home/salo/.cargo/registry/src/index.crates.io-6f17d22bba15001f/ironbar-0.15.1/src/clients/compositor/hyprland.rs:247:4
    |
247 | fn get_workspace_name(name: WorkspaceType) -> String {
    |    ^^^^^^^^^^^^^^^^^^ -------------------

Some errors have detailed explanations: E0308, E0432.
For more information about an error, try `rustc --explain E0308`.
error: could not compile `ironbar` (bin "ironbar") due to 4 previous errors
error: failed to compile `ironbar v0.15.1`, intermediate artifacts can be found at `/tmp/cargo-installuwAhi3`.
To reuse those artifacts with a future compilation, set the environment variable `CARGO_TARGET_DIR` to that path.
```

### Then when building from source it compiles but doesn't start
```
./target/release/ironbar
2024-07-21T01:58:31.046378Z  INFO ironbar: 132: Ironbar version 0.16.0-pre
2024-07-21T01:58:31.046402Z  INFO ironbar: 133: Starting application
2024-07-21T01:58:31.146507Z  INFO ironbar::ipc::server: 39: Starting IPC on /run/user/1000/ironbar-ipc.sock
2024-07-21T01:58:31.150016Z  INFO ironbar::bar: 118: Initializing bar 'soironbatop' on 'eDP-1'
2024-07-21T01:58:31.207793Z ERROR ironbar: 220: 
   0: zbus error
   1: name already taken on the bus

Location:
   src/clients/mod.rs:121

Backtrace omitted. Run with RUST_BACKTRACE=1 environment variable to display it.
Run with RUST_BACKTRACE=full to include source snippets.
2024-07-21T01:58:31.292942Z  WARN ironbar::image::provider: 131: Failed to find image: com.sayonara-player.Sayonara
^C2024-07-21T01:58:35.139859Z  INFO ironbar: 190: Shutting down
```

**System information:**
 - Distro: openSuse tumblweed
- Compositor: labwc
 - Ironbar version: 0.16.0-pre

**Configuration**
> Share your bar configuration and stylesheet as applicable:

<details><summary>Config</summary>

```
name: soironbatop
anchor_to_edges: true
icon_theme: breeze-dark
position: top
height: 20
popup_gap: 0
start:
  - type: label
    name: 'reload-conf'
    label: 
    tooltip: 'Reload bar configuration'
    on_click_left: 'ironbar reload'
  - name: sosep
    label: '|'
    type: label
  - type: custom
    name: soworkspace
    bar:
      - type: box
        widgets:
        - type: label
          label: '1'
          class: ws-lbl
          tooltip: 'Go to workspace 1'
          on_click_left: 'wtype -M win -P F1'
        - type: label
          label: '2'
          class: ws-lbl
          tooltip: 'Go to workspace 2'
          on_click_left: 'wtype -M win -P F2'
        - type: label
          label: '3'
          class: ws-lbl
          tooltip: 'Go to workspace 3'
          on_click_left: 'wtype -M win -P F3'
        - type: label
          label: '4'
          class: ws-lbl
          tooltip: 'Go to workspace 4'
          on_click_left: 'wtype -M win -P F4'
  - name: sosep
    label: '|'
    type: label
  - icon_size: 18
    show_icons: true
    show_names: false
    type: launcher
  - name: sosep
    label: '|'
    type: label
  - name: focused-window
    type: focused
    show_icon: true
    show_title: true
    icon_size: 16
    truncate: end
end:
  - name: sosep
    label: '|'
    type: label
  - music_dir: /media/SDisque/Disque-S/Audio
    player_type: mpris
    format: ''
    show_status_icon: true
    icon_size: 14
    type: music
    truncate:
      max_length: 100
      mode: end
  - format:
      -  {cpu_percent}% | {temp_c:coretemp-Package-id-0}°C
      -  {memory_used} GB ({memory_percent}%)
    type: sys_info
    interval:
      cpu: 1
      memory: 30
      temps: 5
  - name: sosep
    label: '|'
    type: label
  - class: sovolume
    tooltip: Volume
    type: custom
    bar:
      - name: sovolume-btn
        on_click: popup:toggle
        label: '{{so-volumeicon.sh}}'
        type: button
    popup:
      - orientation: v
        type: box
        widgets:
          - label: Volume Settings
            name: header
            type: label
          - name: sovol-box
            type: box
            orientation: h
            widgets:
              - type: box
                orientation: v
                class: vol-box
                widgets:
                  - type: label
                    label: 'Speakers:'
                    class: vol-label
                  - type: label
                    label: 'Microphone:'
                    class: vol-label
              - type: box
                class: vol-box
                orientation: v
                widgets:
                  - class: vol-slider
                    tooltip: Change Volume
                    type: slider
                    orientation: h
                    max: 100
                    on_change: '!iron-volume.sh sink set volume ${0}%'
                    show_label: true
                    length: 150
                    value: |-
                        iron-volume.sh sink get volume
                    step: 5
                  - class: vol-slider
                    tooltip: Change Volume
                    type: slider
                    orientation: h
                    max: 100
                    length: 150
                    step: 5
                    show_label: true
                    value: |-
                        iron-volume.sh source get volume
                    on_change: '!iron-volume.sh source set volume ${0}%'
              - type: box
                orientation: v
                class: vol-box
                widgets:
                  - type: button
                    class: vol-btn
                    label: '{{iron-volume.sh sink get mute}}'
                    on_click: '!iron-volume.sh sink set mute'
                  - type: button
                    class: vol-btn
                    label: '{{iron-volume.sh source get mute}}'
                    on_click: '!iron-volume.sh source set mute'
  - format: '{percentage}%'
    icon_size: 14
    tooltip: 'UPower'
    type: upower
  - max_items: 10
    type: clipboard
  - direction: left_to_right
    type: tray
  - format: '%H:%M — %a %d-%m'
    format_popup: |-
      %H:%M
      %B %d-%m
    type: clock
  - class: power-menu
    tooltip: 'Up: {{30000:uptime | cut -d'','' -f1 | cut -d'' '' -f5-}}'
    type: label
    label: 
    on_click_left: wlogout -r 5 -c 5

```
</details>


<details><summary>Styles</summary>

```css
@define-color color_bg #31363B;
@define-color color_bg_dark grey; /* #1c1c1c;*/
@define-color color_border #424242;
@define-color color_border_active #3daee9; /*rgb(61,174,233)*/
@define-color color_text #FBFBFB;
@define-color color_urgent #8f0a0a;

/* -- base styles -- */

* {
    font-family: Noto Sans Nerd Font, sans-serif;
    font-size: 12px;
    border: none;
    border-radius: 0;
}

box, menubar, button {
    background-color: @color_bg;
    background-image: none;
    box-shadow: none;
    padding: 0;
}

button, label {
    color: @color_text;
}

button:hover {
/*     background-color: @color_bg_dark; */
    background-color: @color_bg;
}

scale trough {
    min-width: 1px;
    min-height: 2px;

}

scale trough highlight  {
    background-color: @color_border_active;
    background-image: none;
    /*background-clip: content-box;*/
}

scale slider {
    background-color: @color_text;
}

#bar {
    border-bottom: 1px solid @color_bg_dark ;
    padding: 2px 5px 2px 5px;
}

.popup {
    border: 1px solid @color_border;
    padding: 1em;
    border-radius: 5px;
}


/* -- clipboard -- */

.clipboard {
    margin-left: 5px;
    /*font-size: 1.1em;*/
    font-size: 1em;
}

.popup-clipboard .item {
    padding-bottom: 0.3em;
    /*border-bottom: 1px solid @color_border;  When bar in in bottom screen*/
    border-top: 1px solid @color_border;
}


/* -- clock -- */

.clock {
    font-weight: bold;
    margin: 0 5px;
}

.popup-clock .calendar-clock {
    color: @color_text;
    font-size: 2em;
    padding-bottom: 0.1em;
}

.popup-clock .calendar {
    background-color: @color_bg;
    color: @color_text;
}

.popup-clock .calendar .header {
    padding-top: 1em;
    border-top: 1px solid @color_border;
    font-size: 1.5em;
}

.popup-clock .calendar:selected {
    background-color: @color_border_active;
}


/* -- launcher -- */

.launcher .item {
    margin: 0 3px;
    padding: 0 3px;
}

.launcher .item:hover {
     background-color: @color_bg_dark;
}

.launcher .ifix examtem:not(.focused):hover {
    background-color: @color_bg_dark;
}

.launcher .open {
    border-bottom: 1px solid @color_text;
}

.launcher .focused {
    border-bottom: 2px solid @color_border_active;
/*     background-color: @color_bg_dark; */
}

.launcher .urgent {
    border-bottom-color: @color_urgent;
}

.popup-launcher {
    padding: 0;
}

.popup-launcher .popup-item:not(:first-child) {
    border-top: 1px solid @color_border;
}


/* -- music -- */

.music {
    margin: 0 5px
}

.music:hover * {
    background-color: @color_bg_dark;
}

.popup-music  {
    border-radius: 5px;
}

.popup-music .album-art {
    margin-right: 1em;
}

.popup-music .icon-box {
    margin-right: 0.4em;
}

.popup-music .title .icon, .popup-music .title .label {
    font-size: 1.7em;
}

.popup-music .controls *:disabled {
    color: @color_border;
}

.popup-music .controls *  {
    font-size: 1.2em;
}

.popup-music .volume .slider slider {
    border-radius: 100%;
    color: @color_text;
}

.popup-music .volume .icon {
    margin-left: 4px;
    font-size: 1.2em;
}

.popup-music .progress .slider slider {
    border-radius: 100%;

}

/* -- script -- */

.script {
    padding-left: 10px;
}


/* -- sys_info -- */

.sysinfo {
    margin: 0 5px;
}

.sysinfo .item {
    margin-left: 2px;
}

/* -- tray -- */

.tray {
    margin: 0 5px 0 5px;
}
.tray .item {
    padding: 2px 3px;
}
.tray .item:hover {
    background-color: rgba(61,174,233, 0.5);
}

/* -- workspaces -- */

.workspaces .item.focused {
    box-shadow: inset 0 -3px;
    background-color: @color_bg_dark;
}

.workspaces .item:hover {
    box-shadow: inset 0 -3px;
}

/* -- custom: power menu -- */
.power-menu {
    margin: 0 5px;
}
#power-btn {
    margin: 0 5px;
}

.popup-power-menu #header {
    font-size: 1.4em;
    padding-bottom: 0.4em;
    margin-bottom: 0.6em;
    border-bottom: 1px solid @color_border;
}

.popup-power-menu .power-btn {
    border: 1px solid @color_border;
    padding: 0.6em 1em;
}

.popup-power-menu .power-btn:hover {
    border: 1px solid @color_border_active;
}

.popup-power-menu #buttons > *:nth-child(1) .power-btn {
    margin-right: 1em;
}

#sosep {
    padding: 0;
    margin: 0 0 2px 0;
    color: darkgrey;
}

#focused-window {
/*     border: 1px solid @color_bg_darkborder-radius: 5px;; */
/* border-radius: 5px; */
    padding: 2px 4px;
    font-weight: bold;
}

.upower {
    margin: 0 5px;
}

.popup-upower {
    border-radius: 5px;
}

#sovolume-btn {
    padding: 0;
    margin: 0 5px;
}

.popup-sovolume #header {
    font-size: 1.4em;
    color: @color_text;
    padding-bottom: 0.4em;
    margin-bottom: 0.6em;
    border-bottom: 1px solid @color_border;
}

.popup-sovolume {
    padding: 1em;
}

#sovol-box {
    padding: 0.5em;
/*     border: 1px solid red; */

}

.popup-sovolume .vol-box {
     margin: 0.2em;
/*    border: 1px solid green; */
}

.popup-sovolume .vol-btn {
    font-size: 1.2em;
    font-weight: bold;
    border: 1px solid @color_border;
    padding: 0.5em 2em;
    margin: 0.5em;
    border-radius: 5px;
}

.popup-sovolume .vol-btn:hover, .vol-slider:hover {
    border: 1px solid @color_border_active;
}

.vol-slider {
    border: 1px solid @color_bg;
     padding: 0 2em 1.5em 2em ;
    margin: 0.5em;
    border-radius: 5px;
}

/*.vol-slider:hover {
    border: 1px solid @color_border_active;
}*/

.popup-sovolume .vol-slider slider {
    border-radius: 100%;
    margin-left: 5px;
    margin-right: 5px;
}

.vol-label {
    font-size: 1.2em;
    /*border: 1px solid @color_border;*/
    padding: 0.5em 2em;
    margin: 0.5em;
    border-radius: 5px;
}

#reload-conf {
    font-weight: bold;
    margin: 0 5px;
}

.ws-lbl {
    margin: 0 5px;
    /*border: 1px solid @color_border_active;
    border-radius: 100%;*/
}

```
</details>

I tried to install `ironbar` again using `cargo`, but it doesn't compile, same as before, guess it's still on `0.15.1`
```
error[E0432]: unresolved import `nix::fcntl::OFlag`
 --> /home/salo/.cargo/registry/src/index.crates.io-6f17d22bba15001f/ironbar-0.15.1/src/clients/wayland/wlr_data_control/offer.rs:3:5
  |
3 | use nix::fcntl::OFlag;
  |     ^^^^^^^^^^^^^^^^^ no `OFlag` in `fcntl`

error[E0432]: unresolved import `nix::unistd::pipe2`
    --> /home/salo/.cargo/registry/src/index.crates.io-6f17d22bba15001f/ironbar-0.15.1/src/clients/wayland/wlr_data_control/offer.rs:4:26
     |
4    | use nix::unistd::{close, pipe2};
     |                          ^^^^^
     |                          |
     |                          no `pipe2` in `unistd`
     |                          help: a similar name exists in the module: `pipe`
     |
note: found an item that was configured out
    --> /home/salo/.cargo/registry/src/index.crates.io-6f17d22bba15001f/nix-0.27.1/src/unistd.rs:1233:8
     |
1233 | pub fn pipe2(flags: OFlag) -> Result<(RawFd, RawFd)> {
     |        ^^^^^
     = note: the item is gated behind the `fs` feature

error[E0432]: unresolved imports `nix::fcntl::fcntl`, `nix::fcntl::F_GETPIPE_SZ`, `nix::fcntl::F_SETPIPE_SZ`
   --> /home/salo/.cargo/registry/src/index.crates.io-6f17d22bba15001f/ironbar-0.15.1/src/clients/wayland/wlr_data_control/mod.rs:13:18
    |
13  | use nix::fcntl::{fcntl, F_GETPIPE_SZ, F_SETPIPE_SZ};
    |                  ^^^^^  ^^^^^^^^^^^^  ^^^^^^^^^^^^ no `F_SETPIPE_SZ` in `fcntl`
    |                  |      |
    |                  |      no `F_GETPIPE_SZ` in `fcntl`
    |                  no `fcntl` in `fcntl`
    |
    = help: consider importing this module instead:
            nix::fcntl
note: found an item that was configured out
   --> /home/salo/.cargo/registry/src/index.crates.io-6f17d22bba15001f/nix-0.27.1/src/fcntl.rs:509:8
    |
509 | pub fn fcntl(fd: RawFd, arg: FcntlArg) -> Result<c_int> {
    |        ^^^^^
    = note: the item is gated behind the `fs` feature

error[E0308]: mismatched types
   --> /home/salo/.cargo/registry/src/index.crates.io-6f17d22bba15001f/ironbar-0.15.1/src/clients/compositor/hyprland.rs:156:51
    |
156 |                     let name = get_workspace_name(workspace_type);
    |                                ------------------ ^^^^^^^^^^^^^^ expected `WorkspaceType`, found `WorkspaceDestroyedEventData`
    |                                |
    |                                arguments to this function are incorrect
    |
note: function defined here
   --> /home/salo/.cargo/registry/src/index.crates.io-6f17d22bba15001f/ironbar-0.15.1/src/clients/compositor/hyprland.rs:247:4
    |
247 | fn get_workspace_name(name: WorkspaceType) -> String {
    |    ^^^^^^^^^^^^^^^^^^ -------------------

Some errors have detailed explanations: E0308, E0432.
For more information about an error, try `rustc --explain E0308`.
error: could not compile `ironbar` (bin "ironbar") due to 4 previous errors
error: failed to compile `ironbar v0.15.1`, intermediate artifacts can be found at `/tmp/cargo-installd9FUG2`.
To reuse those artifacts with a future compilation, set the environment variable `CARGO_TARGET_DIR` to that path.
```

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -149,6 +149,7 @@ sysinfo = { version = "0.29.11", optional = true }
 
 # tray
 system-tray = { version = "0.2.0", optional = true }
+system-tray = { version = "0.3.0", optional = true }
 
 # upower
 upower_dbus = { version = "0.3.2", optional = true }

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -68,7 +68,7 @@ notifications = ["zbus"]
 
 sys_info = ["sysinfo", "regex"]
 
-tray = ["system-tray"]
+tray = ["system-tray", "png"]
 
 upower = ["upower_dbus", "zbus", "futures-lite"]
 

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -148,8 +148,8 @@ futures-signals = { version = "0.3.34", optional = true }
 sysinfo = { version = "0.29.11", optional = true }
 
 # tray
-system-tray = { version = "0.2.0", optional = true }
 system-tray = { version = "0.3.0", optional = true }
+png = { version = "0.17.14", optional = true }
 
 # upower
 upower_dbus = { version = "0.3.2", optional = true }

--- a/src/clients/mod.rs
+++ b/src/clients/mod.rs
@@ -1,4 +1,4 @@
-use crate::{await_sync, Ironbar};
+use crate::await_sync;
 use color_eyre::Result;
 use std::path::Path;
 use std::rc::Rc;
@@ -151,9 +151,7 @@ impl Clients {
         let client = match &self.tray {
             Some(client) => client.clone(),
             None => {
-                let service_name = format!("{}-{}", env!("CARGO_CRATE_NAME"), Ironbar::unique_id());
-
-                let client = await_sync(async { tray::Client::new(&service_name).await })?;
+                let client = await_sync(async { tray::Client::new().await })?;
                 let client = Arc::new(client);
                 self.tray.replace(client.clone());
                 client

--- a/src/main.rs
+++ b/src/main.rs
@@ -464,7 +464,7 @@ where
 /// Blocks on a `Future` until it resolves.
 ///
 /// This is not an `async` operation
-/// so can be used outside of an async function.
+/// so can be used outside an async function.
 ///
 /// Use sparingly, as this risks blocking the UI thread!
 /// Prefer async functions wherever possible.

--- a/src/modules/tray/diff.rs
+++ b/src/modules/tray/diff.rs
@@ -18,7 +18,9 @@ pub struct MenuItemDiff {
     /// True if the item is visible in the menu.
     pub visible: Option<bool>,
     /// Icon name of the item, following the freedesktop.org icon spec.
-    // pub icon_name: Option<Option<String>>,
+    pub icon_name: Option<Option<String>>,
+    /// PNG icon data.
+    pub icon_data: Option<Option<Vec<u8>>>,
     /// Describe the current state of a "togglable" item. Can be one of:
     ///   - Some(true): on
     ///   - Some(false): off
@@ -52,7 +54,8 @@ impl MenuItemDiff {
             label: diff!(&label),
             enabled: diff!(enabled),
             visible: diff!(visible),
-            // icon_name: diff!(&icon_name),
+            icon_name: diff!(&icon_name),
+            icon_data: diff!(&icon_data),
             toggle_state: diff!(toggle_state),
             submenu: get_diffs(&old.submenu, &new.submenu),
         }

--- a/src/modules/tray/icon.rs
+++ b/src/modules/tray/icon.rs
@@ -7,6 +7,7 @@ use gtk::ffi::gtk_icon_theme_get_search_path;
 use gtk::gdk_pixbuf::{Colorspace, InterpType, Pixbuf};
 use gtk::prelude::IconThemeExt;
 use gtk::{IconLookupFlags, IconTheme, Image};
+use png::ColorType;
 use std::collections::HashSet;
 use std::ffi::CStr;
 use std::os::raw::{c_char, c_int};
@@ -124,4 +125,37 @@ fn get_image_from_pixmap(item: &TrayMenu, size: u32) -> Result<Image> {
     let image = Image::new();
     ImageProvider::create_and_load_surface(&pixbuf, &image)?;
     Ok(image)
+}
+
+pub struct PngData<'a>(pub &'a [u8]);
+impl TryFrom<PngData<'_>> for Image {
+    type Error = Report;
+
+    fn try_from(value: PngData) -> std::result::Result<Self, Self::Error> {
+        let data = value.0;
+
+        let decoder = png::Decoder::new(data);
+        let mut reader = decoder.read_info()?;
+        let mut buf = vec![0; reader.output_buffer_size()];
+
+        let info = reader.next_frame(&mut buf)?;
+        let bytes = glib::Bytes::from(&buf[..info.buffer_size()]);
+
+        let has_alpha = matches!(info.color_type, ColorType::Rgba | ColorType::GrayscaleAlpha);
+        let row_stride_multiplier = if has_alpha { 4 } else { 3 };
+
+        let pixbuf = Pixbuf::from_bytes(
+            &bytes,
+            Colorspace::Rgb,
+            has_alpha,
+            info.bit_depth as i32,
+            info.width as i32,
+            info.height as i32,
+            (info.width * row_stride_multiplier) as i32,
+        );
+
+        let image = Image::new();
+        ImageProvider::create_and_load_surface(&pixbuf, &image)?;
+        Ok(image)
+    }
 }

--- a/src/modules/tray/interface.rs
+++ b/src/modules/tray/interface.rs
@@ -343,9 +343,13 @@ impl TrayMenuItem {
         }
 
         // TODO: Image support
-        // if let Some(icon_name) = diff.icon_name {
-        //
-        // }
+        if let Some(_icon_name) = diff.icon_name {
+            warn!("received unimplemented menu icon update");
+        }
+
+        if let Some(_icon_data) = diff.icon_data {
+            warn!("received unimplemented menu icon update");
+        }
 
         if let Some(enabled) = diff.enabled {
             match &self.widget {

--- a/src/modules/tray/interface.rs
+++ b/src/modules/tray/interface.rs
@@ -1,13 +1,19 @@
 use super::diff::{Diff, MenuItemDiff};
+use crate::image::ImageProvider;
+use crate::modules::tray::icon::PngData;
 use crate::{spawn, try_send};
 use glib::Propagation;
 use gtk::prelude::*;
-use gtk::{CheckMenuItem, Image, Label, Menu, MenuItem, SeparatorMenuItem};
+use gtk::{
+    CheckMenuItem, Container, IconTheme, Image, Label, Menu, MenuItem, Orientation,
+    SeparatorMenuItem,
+};
 use std::collections::HashMap;
 use system_tray::client::ActivateRequest;
 use system_tray::item::{IconPixmap, StatusNotifierItem};
 use system_tray::menu::{MenuItem as MenuItemInfo, MenuType, ToggleState, ToggleType};
 use tokio::sync::mpsc;
+use tracing::{error, warn};
 
 /// Calls a method on the underlying widget,
 /// passing in a single argument.
@@ -214,6 +220,37 @@ enum TrayMenuWidget {
     Checkbox(CheckMenuItem),
 }
 
+fn setup_item<W>(widget: &W, info: &MenuItemInfo)
+where
+    W: IsA<MenuItem> + IsA<Container>,
+{
+    let container = gtk::Box::new(Orientation::Horizontal, 10);
+    widget.add(&container);
+
+    if let Some(icon) = &info.icon_name {
+        // TODO: Get theme here
+        let image = Image::new();
+        match ImageProvider::parse(icon, &IconTheme::new(), true, 24)
+            .map(|provider| provider.load_into_image(image.clone()))
+        {
+            Some(Ok(())) => container.add(&image),
+            _ => warn!("Failed to load icon: {icon}"),
+        }
+    }
+
+    if let Some(icon_data) = &info.icon_data {
+        match Image::try_from(PngData(icon_data.as_slice())) {
+            Ok(image) => container.add(&image),
+            Err(err) => error!("{err:?}"),
+        };
+    }
+
+    let label = Label::new(info.label.as_deref());
+    container.add(&label);
+
+    container.show_all();
+}
+
 impl TrayMenuItem {
     fn new(info: &MenuItemInfo, tx: mpsc::Sender<i32>) -> Self {
         let mut submenu = HashMap::new();
@@ -234,7 +271,9 @@ impl TrayMenuItem {
         }
 
         let widget = match (info.menu_type, info.toggle_type) {
-            (MenuType::Separator, _) => TrayMenuWidget::Separator(SeparatorMenuItem::new()),
+            (MenuType::Separator, _) => {
+                TrayMenuWidget::Separator(SeparatorMenuItem::builder().visible(true).build())
+            }
             (MenuType::Standard, ToggleType::Checkmark) => {
                 let widget = CheckMenuItem::builder()
                     .visible(info.visible)
@@ -242,10 +281,7 @@ impl TrayMenuItem {
                     .active(info.toggle_state == ToggleState::On)
                     .build();
 
-                if let Some(label) = &info.label {
-                    widget.set_label(label);
-                }
-
+                setup_item(&widget, info);
                 add_submenu!(menu, widget);
 
                 {
@@ -266,10 +302,7 @@ impl TrayMenuItem {
                     .sensitive(info.enabled)
                     .build();
 
-                if let Some(label) = &info.label {
-                    widget.set_label(label);
-                }
-
+                setup_item(&widget, info);
                 add_submenu!(menu, widget);
 
                 {

--- a/src/modules/tray/mod.rs
+++ b/src/modules/tray/mod.rs
@@ -16,7 +16,7 @@ use std::collections::HashMap;
 use system_tray::client::Event;
 use system_tray::client::{ActivateRequest, UpdateEvent};
 use tokio::sync::mpsc;
-use tracing::{debug, error, warn};
+use tracing::{debug, error, trace, warn};
 
 #[derive(Debug, Deserialize, Clone)]
 #[cfg_attr(feature = "schema", derive(schemars::JsonSchema))]
@@ -198,7 +198,8 @@ fn on_update(
             menus.insert(address.into(), menu_item);
         }
         Event::Update(address, update) => {
-            debug!("Received tray update for '{address}': {update:?}");
+            debug!("Received tray update for '{address}'");
+            trace!("Tray update for '{address}: {update:?}'");
 
             let Some(menu_item) = menus.get_mut(address.as_str()) else {
                 error!("Attempted to update menu at '{address}' but could not find it");

--- a/src/modules/tray/mod.rs
+++ b/src/modules/tray/mod.rs
@@ -211,13 +211,12 @@ fn on_update(
                 }
                 UpdateEvent::Icon(icon) => {
                     if icon.as_ref() != menu_item.icon_name() {
+                        menu_item.set_icon_name(icon);
                         match icon::get_image(menu_item, icon_theme, icon_size, prefer_icons) {
                             Ok(image) => menu_item.set_image(&image),
                             Err(_) => menu_item.show_label(),
                         };
                     }
-
-                    menu_item.set_icon_name(icon);
                 }
                 UpdateEvent::OverlayIcon(_icon) => {
                     warn!("received unimplemented NewOverlayIcon event");

